objects/src/accounts/account_id.rs

```diff
@@ -1,53 +1,65 @@
+///  - 00 - regular account with updatable code.
+///  - 01 - regular account with immutable code.
+///  - 10 - fungible asset faucet with immutable code.
+///  - 11 - non-fungible asset faucet with immutable code.
+/// - The third most significant bit of the ID specifies whether the account data is stored on-chain:
+///  - 0 - full account data is stored on-chain.
+///  - 1 - only the account hash is stored on-chain which serves as a commitment to the account state.
+/// As such the three most significant bits fully describes the type of the account.
 ///
 /// Additionally, account IDs have the following properties
-/// - For fungible asset faucets account IDs are guaranteed to start with ONE.
-/// - For regular accounts, the last 3 bytes of the ID are guaranteed to be all ZEROs.
+/// - Faucet account IDs are guaranteed to start with ONE.
+//  - Regular account IDs are guaranteed to start with ZERO.
 #[derive(Debug, Copy, Clone, Eq, PartialEq)]
-pub struct AccountId([Felt; 3]);
+pub struct AccountId(Felt);
 
 impl AccountId {
     // CONSTANTS
     // --------------------------------------------------------------------------------------------
-    pub const FUNGIBLE_FAUCET_TAG: u32 = 0;
-    pub const NON_FUNGIBLE_FAUCET_TAG: u32 = 1 << 31;
+    pub const FUNGIBLE_FAUCET_TAG: u64 = 0b10;
+    pub const NON_FUNGIBLE_FAUCET_TAG: u64 = 0b11;
+    pub const REGULAR_ACCOUNT_UPDATABLE_CODE_TAG: u64 = 0b00;
+    pub const REGULAR_ACCOUNT_IMMUTABLE_CODE_TAG: u64 = 0b01;
+    pub const ON_CHAIN_ACCOUNT_TAG: u64 = 1;
+    pub const OFF_CHAIN_ACCOUNT_TAG: u64 = 0;
 
-    /// Specifies a minimum number of trailing zeros for a valid account ID. Thus, all valid
-    /// account IDs have the last 3 bytes set to zeros.
-    pub const MIN_TRAILING_ZEROS: u32 = 24;
+    /// Specifies a minimum number of trailing zeros required in the last element of the seed digest.
+    pub const REGULAR_ACCOUNT_SEED_DIGEST_MIN_TRAILING_ZEROS: u32 = 24;
+    pub const FAUCET_SEED_DIGEST_MIN_TRAILING_ZEROS: u32 = 32;
+
+    /// Specifies a minimum number of ones for a valid account ID.
+    pub const MIN_ACCOUNT_ONES: u32 = 5;
 
     // CONSTRUCTOR
     // --------------------------------------------------------------------------------------------
     /// Returns a new account ID derived from the specified seed.
     ///
-    /// The account ID is computed by hashing the seed and using 3 elements of the result to form
-    /// the ID. Specifically we are take elements 0, 1, and 3, omitting element 2. We omit element
-    /// 2 because unlike elements 0 and 3, it has no special structure which we need to carry over
-    /// into the derived account ID. Element 1 could have been omitted just as well.
+    /// The account ID is computed by hashing the seed and using 1 element of the result to form
+    /// the ID. Specifically we take element 0. We also require that the last element of the seed
+    /// digest has at least `24` trailing zeros.
```

Comment:
Should we clarify there that depending on the account type the number of trailing zeros is between 24 and 32?